### DISCUSSION TITLE
select() with nested relations is getting slow (timeout)

### DISCUSSION THREAD

**Discussion by patrykgulas** [supabase/supabase#10548] [Questions] [answered]
I have the following simplified tables:

Post
| id | title |
|----|--------|
| POST_1  | Post 1 |
| POST_2  | Post 2 |

Comment
| id | commentText | postId | typeId |
|----|--------|-----|----|
| COMMENT_1  | Comment 1 | POST_1 | TYPE_2 |
| COMMENT_2  | Comment 2 | POST_1 | TYPE_1 |

CommentType
| id | name |
|----|--------|
| TYPE_1  | CommentType 1 |
| TYPE_2  | CommentType 2 |

I'm querying the data on the client side with:

```
supabase.from("Post").select(`
 id,
 title,
 comments:Comment (
    type:CommentType (
      name
    )
  )
`, {
  count: "exact",
}).order("title").range(0, 19);
```

The query is resolving fast with low ranges, like 0 - 19, 20 - 39 but is getting slower for bigger numbers like 100 - 119, 1000 - 1019. I wonder what can be the cause for such behaviour 🤔 I'm trying to implement pagination with range method.

The query returns an error `FetchError: fetch failed` or times out.

Thanks for all the help!

**Comment by GaryAustin1** [comment, marked as answer]
Normally to use range with pagination you need to use .order() to establish a sort column.
If you don't the rows are random and may even be repeated as you step thru your pages.
I don't know if that would have anything to do with performance though.
You might also try turning RLS on and off and see if that is part of your performance issue as you are accessing 3 tables.

**Reply by patrykgulas** [reply]
Thanks for the reply! I added `.order()` to `.select()` but it seems to not have an impact here 🤔 I also have RLS disabled on all tables at this moment.

I forgot to add that I use `count: "exact"` so that probably can also affect performance.

It looks like the higher `from/to` in `.range()` is, the slower response is 🤔

**Reply by GaryAustin1** [reply]
I don't understand why a higher from to would matter in Postgres, but Postgres can be "magical" at times.  
I believe all the instances now support PostgREST 10 which has an explain option you can use the supabase-js to get query info.  That might be useful to use here.

https://github.com/supabase/postgrest-js/blob/master/src/PostgrestTransformBuilder.ts#L174

**Reply by steve-chavez** [reply]
> I don't understand why a higher from to would matter in Postgres

This is a general problem on SQL databases([ref](https://stackoverflow.com/questions/4481388/why-does-mysql-higher-limit-offset-slow-the-query-down)), higher OFFSETs have to scan more rows than lower ones.

**Edit**: @GaryAustin1 This is actually mentioned on the PostgreSQL docs:

> The rows skipped by an OFFSET clause still have to be computed inside the server; therefore a large OFFSET might be inefficient.

https://www.postgresql.org/docs/current/queries-limit.html

@patrykgulas An option might be using keyset pagination, check [this article](https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/).

**Reply by patrykgulas** [reply]
@steve-chavez thanks for the resources, very helpful. It seems to be that case, I need to think which option to choose then as I want to be able to pick exact page.

**Reply by e0** [reply]
I was rewriting some things from `postgres.js` to `supabase-js` (to be able to use Vercel Edge Functions) and noticed the same problem. I don't think this specific problem is due to PostgreSQL. With `postgres.js` before, the response time was pretty consistent regardless of the page. With `supabase-js`, the response time goes from ~300ms to ~4s when going from offset 0 to offset 11000.

As a workaround, I'm only fetching the main table first and then fetching the joined table using something like `in('id', [...mainTableIds])`. There is a slight increase in the response time for 0 offset but now all of the pages provide consistent response times. This is fine for now but it would be nice if we can look more into this. Let me know if there is anything I can do to help.

**Reply by steve-chavez** [reply]
@eo Sure, thanks a lot for reporting.

> With postgres.js before, the response time was pretty consistent regardless of the page. With supabase-js, the response time goes from ~300ms to ~4s when going from offset 0 to offset 11000.

Can you get a hold of the `postgres.js` generated query? I'm unfamiliar with it, so not sure if they might be abstracting `offset` and generating another statement instead.

**Reply by e0** [reply]
> Can you get a hold of the `postgres.js` generated query? I'm unfamiliar with it, so not sure if they might be abstracting `offset` and generating another statement instead.

Sorry, it's a bit inconvenient for me to get the `postgres.js` query atm. I ran then following `SQL` and it takes ~200ms. The same thing using `range` after `.from('summaries').select('*, timeseries ( daily_condensed )', { count: 'exact' })` takes ~4s. I tried both with and without `{ count: 'exact' }` and it didn't make much of a difference.

```sql
select s.*, t.daily_condensed
from summaries as s
join timeseries as t
on s.symbol = t.symbol
offset 11000
limit 30
```

**Reply by GaryAustin1** [reply]
Do you happen to now have a user or RLS involved when using supabase-js versus not in the postgrest-js case.  
At a minimum toggle RLS off and see if that has an impact.  If all the rows have to be processed thru RLS could be a significant hit.

**Reply by e0** [reply]
Ah that's probably it then @GaryAustin1 . For `postgres.js` I had to do a custom auth check first and then used the service role to fetch the data. Same with raw SQL of course. Again, I'm not able to verify this atm but if I get the chance I will try to remember to add an update here.

Do you know if there is a Postgrest issue for degraded performance when combining RLS, related tables, and pagination? Would like to understand more about why RLS has an affect here.

**Reply by GaryAustin1** [reply]
It could be because it has to scan each of those rows being skipped to know if to count them and has to do RLS on each.

You can test RLS on your SQL.   
This is a bit overkill, but should work:
https://supabase.com/docs/guides/auth/row-level-security#testing-policies
